Add tt external modules doc (#3016)
tarantool/doc: doc/reference/tt_cli/external_modules.rst

Extending the tt functionality
==============================

The ``tt`` utility implements the modular architecture: its :doc:`commands <commands>`

> are implemented in separate modules. When you run ``tt`` with a command, the

Reviewer: ```suggestion
The ``tt`` utility implements a modular architecture: its :doc:`commands <commands>`
are, in fact, separate modules. When you run ``tt`` with a command, the
```
2 times "implemented". Feel free to use another wording

corresponding module is executed with the given arguments.

> Modular architecture enables the option to extend the ``tt`` functionality with

Reviewer: ```suggestion
The modular architecture enables the option to extend the ``tt`` functionality with
```

**external modules** (as opposed to **internal modules** that implement built-in
commands). Simply said, you can write any code you want to execute
from ``tt``, pack it into an executable, and run it with a ``tt`` command:

..  code-block:: bash

    tt my-module-name my-args

> The command that executes a module is the same as the name of its executable.

Reviewer: ```suggestion
The name of the command that executes a module is the same as the name of the module's executable.
```

Module description and help
---------------------------

Executables that implement external ``tt`` modules must have two flags:

-   ``--description`` -- print a short description of the module. It is shown alongside
    the command in the ``tt`` help.

> -   ``--help`` -- display help. It is shown when ``tt <module-name> help`` is called.

Reviewer 2: Actually, yes, but no) We pass all options and arguments that were specified by the user into the module (and `tt EXT_MODULE_NAME --help` will work fine), but the `help` module will also output help from the `tt help EXT_MODULE_NAME` external module if `--help` is implemented in an external module.

> -   ``--help`` -- display help. It is shown when ``tt help <module_name>`` is called.

Reviewer: ```suggestion
-   ``--description`` -- print a short description of the module. The description is shown alongside
    the command in the ``tt`` help.
-   ``--help`` -- display help. The help message is shown when ``tt help <module_name>`` is called.
```
It is unclear what the pronoun "it" refers to, the flag or the message

Location
--------

External modules must be located in the modules directory specified in the
:ref:`configuration file <tt-config_file>`:

    ..  code:: yaml

        tt:
            modules:

>                 directory: path/to/modules/dir

Reviewer: ```suggestion
    ..  code:: yaml

        tt:
          modules:
            directory: path/to/modules/dir
```
The suggested syntax for YAML files is to use 2 spaces for indentation. This is not a strict rule, but all the other YAML examples in our doc have 2-space indents.

To check if a module is available in ``tt``, call ``tt help``.

> It will show available external modules in the ``EXTERNAL COMMANDS`` section together

Reviewer: ```suggestion
It will show the available external modules in the ``EXTERNAL COMMANDS`` section together
```

with their descriptions.

Overloading built-in commands
-----------------------------

External modules can overload built-in ``tt`` commands.
If you want to change the behavior of a built-in command, create an external

> modules with the same name and your own implementation.

Reviewer: ```suggestion
module with the same name and your own implementation.
```

When ``tt`` sees two modules -- an external and an internal one -- with the same name,
it will use the external module by default.

> For example, if you want to get the information about your Tarantool application
> from ``tt``, write the corresponding external module ``version``. The ``tt version``

Reviewer: ```suggestion
For example, if you want to get the information from ``tt`` about your Tarantool application, write an external module with the name ``version``. The ``tt version``
```
- Consider changing object order. Otherwise, it can be read as "your Tarantool application from `tt`", which is confusing.
- It's not clear what the `version` module "corresponds to."

Author: Thanks, rephrased
